File: netforce_ui_web/ui_params.js
var rpc=require("./rpc");

var _ui_params=null;

module.exports={
    load_ui_params: function(modules,cb) {
        var ctx={modules:modules};
        rpc.execute("ui.params","load_ui_params",[],{context:ctx},function(err,data) {
            if (err) {
                cb(err);
                return;
            }
            _ui_params=data;
            cb();
        });
    },

    get_layout: function(name) {
        if (!_ui_params) throw "UI params not found";
        var l=_ui_params.layouts[name];
        if (!l) throw "Layout not found: "+name;
        return l;
    },

    find_layout: function(conds) {
        if (!_ui_params) throw "UI params not found";
        var layouts=_ui_params.layouts;
        var found=null;
        for (var n in layouts) {
            var l=layouts[n];
            if (l.model!=conds.model) continue;
            if (conds.type && l.type!=conds.type) continue;
            found=l;
        }
        return found;
    },

    get_model: function(model) {
        if (!_ui_params) throw "UI params not found";
        var m=_ui_params.models[model];
        if (!m) throw "Model not found: "+model;
        return m;
    },

    get_field: function(model,name) {
        var m=this.get_model(model);
        f=m.fields[name];
        if (!f) throw "Field not found: "+model+"."+name;
        return f;
    },
}

File: netforce_general/netforce_general/models/ui_params.py
# Copyright (c) 2012-2015 Netforce Co. Ltd.
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
# OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
# OR OTHER DEALINGS IN THE SOFTWARE.

from netforce.model import Model, fields
from netforce import action
from netforce import layout
from netforce import model
from netforce import get_module_version_name,get_module_version_code

class UIParams(Model):
    _name = "ui.params"
    _store = False

    def get_version(self,context={}):
        return {
            "version_name": get_module_version_name(),
            "version_code": get_module_version_code(),
        }
    
    def load_ui_params(self,context={}):
        actions=action.actions_to_json(modules=context.get("modules"))
        layouts=layout.layouts_to_json(modules=context.get("modules"),mobile_only=context.get("mobile_only"))
        models=model.models_to_json()
        return {
            "version_name": get_module_version_name(),
            "version_code": get_module_version_code(),
            "actions": actions,
            "layouts": layouts,
            "models": models,
        }

UIParams.register()
